### PR TITLE
*: support sample for compressed files for adjustment

### PR DESCRIPTION
**Pull request by lichunzhu**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/38514

Problem Summary:

### What is changed and how it works?
* Support computing sample compressed ratio for lightning compressed files.
* Use compress ratio to compute file size for precheck and row id allocating.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- None

Documentation

- None

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/mydump/loader.go [hunk 1]
```diff
@@ -470,6 +476,15 @@ func (s *mdLoaderSetup) constructFileInfo(ctx context.Context, path string, size
 	case SourceTypeViewSchema:
 		s.viewSchemas = append(s.viewSchemas, info)
 	case SourceTypeSQL, SourceTypeCSV, SourceTypeParquet:
+		if info.FileMeta.Compression != CompressionNone {
+			compressRatio, err2 := SampleFileCompressRatio(ctx, info.FileMeta, s.loader.GetStore())
+			if err2 != nil {
+				logger.Error("[loader] fail to calculate data file compress compress ratio",
```
Comment (buchuitoudegou) [inline br/pkg/lightning/mydump/loader.go:482]:
```suggestion
				logger.Error("[loader] fail to calculate data file compress ratio",
```

File: br/pkg/lightning/mydump/loader.go [hunk 2]
```diff
@@ -648,3 +663,75 @@ func (l *MDLoader) GetDatabases() []*MDDatabaseMeta {
...
+	defer reader.Close()
+
+	compressReader, err := storage.NewLimitedInterceptReader(reader, compressType, offset)
+	if err != nil {
+		return 0, 0, errors.Trace(err)
+	}
+
+	readBytes := func() error {
+		n, err2 := compressReader.Read(bytes)
+		if err2 != nil && !strings.Contains(err2.Error(), "EOF") {
```
Comment (buchuitoudegou) [inline br/pkg/lightning/mydump/loader.go:686]: Can we use `errors.Cause(err2) != io.EOF`?

File: br/pkg/lightning/mydump/loader.go [hunk 3]
```diff
@@ -648,3 +663,75 @@ func (l *MDLoader) GetDatabases() []*MDDatabaseMeta {
...
+			return err2
+		}
+		tot += n
+		return err2
+	}
+
+	if offset == 0 {
+		err = readBytes()
+		if err != nil && !strings.Contains(err.Error(), "EOF") {
```
Comment (buchuitoudegou) [inline br/pkg/lightning/mydump/loader.go:695]: ditto

File: br/pkg/lightning/mydump/loader.go [hunk 4]
```diff
@@ -648,3 +663,75 @@ func (l *MDLoader) GetDatabases() []*MDDatabaseMeta {
...
+	if err != nil {
+		return 0, 0, errors.Trace(err)
+	}
+
+	readBytes := func() error {
+		n, err2 := compressReader.Read(bytes)
+		if err2 != nil && !strings.Contains(err2.Error(), "EOF") {
+			return err2
+		}
+		tot += n
```
Comment (buchuitoudegou) [inline br/pkg/lightning/mydump/loader.go:689]: Is `n` equal to `sampleCompressedFileSize`?

File: br/pkg/lightning/mydump/loader.go [hunk 5]
```diff
@@ -648,3 +663,75 @@ func (l *MDLoader) GetDatabases() []*MDDatabaseMeta {
...
+	if err != nil {
+		return 0, err
+	}
+	// read first time, aims to find a valid end pos in compressed file
+	_, pos, err := calculateFileBytes(ctx, fileMeta.Path, compressType, store, 0)
+	if err != nil {
+		return 0, err
+	}
+	// read second time, original reader ends at first time's valid pos, compute sample data compress ratio
+	tot, pos, err := calculateFileBytes(ctx, fileMeta.Path, compressType, store, pos)
```
Comment (buchuitoudegou) [inline br/pkg/lightning/mydump/loader.go:732]: I don't really understand why we need to read twice here? Would you elaborate it in the PR or in the comment?

File: br/pkg/lightning/mydump/region.go [hunk 6]
```diff
@@ -303,11 +301,8 @@ func MakeSourceFileRegion(
 	rowIDMax := fileSize / divisor
 	// for compressed files, suggest the compress ratio is 1% to calculate the rowIDMax.
 	// set fileSize to INF to make sure compressed files can be read until EOF. Because we can't get the exact size of the compressed files.
-	// TODO: update progress bar calculation for compressed files.
 	if fi.FileMeta.Compression != CompressionNone {
-		// FIXME: this is not accurate. Need sample ratio in the future and use sampled ratio to compute rowIDMax
-		//  currently we use 500 here. It's a relatively large value for most data.
-		rowIDMax = fileSize * compressDataRatio / divisor
+		rowIDMax = fi.FileMeta.RealSize * CompressSizeFactor / divisor
```
Comment (buchuitoudegou) [inline br/pkg/lightning/mydump/region.go:305]:
```
info.FileMeta.RealSize = int64(compressRatio * float64(info.FileMeta.FileSize))
```
The real size is the size of the de-compressed file, so we don't need to multiply the CompressSizeFactor anymore?

File: br/pkg/lightning/mydump/loader.go (hunk 2 above)
Comment (lichunzhu) [reply br/pkg/lightning/mydump/loader.go:686]: I'm not sure whether `S3Reader` or `GCSReader` can correctly return `io.EOF`. I check their code and it sames that they can. I will change this.

File: br/pkg/lightning/mydump/loader.go (hunk 5 above)
Comment (lichunzhu) [reply br/pkg/lightning/mydump/loader.go:732]: As the above comment saying:
We read first time aiming to find a valid compressed file offset. If we continue read now, the below file reader may process more data and buffer them in memory.
So we use a second reading to limit the below file reader only read n bytes(n is the valid position we find in the first reading). Then we read all the data out to calculate the compress ratio.

File: br/pkg/lightning/mydump/region.go (hunk 6 above)
Comment (buchuitoudegou) [reply br/pkg/lightning/mydump/region.go:308]: lgtm

File: br/pkg/lightning/mydump/loader.go (hunk 5 above)
Comment (buchuitoudegou) [reply br/pkg/lightning/mydump/loader.go:738]: okay, I finally understand what you mean by reading the compression reader code😢
1. find a valid offset n
2. read n byte of data from the compressed file, then calculate the uncompressed data (m bytes)
3. compress ratio: m/n

Am I right?

**Review by buchuitoudegou — APPROVED** [review]

**Comment by lichunzhu** [comment]
@dsdashun PTAL

**Review by dsdashun — APPROVED** [review]

**Pull request merged**